pkg/assets/create/creater.go

```diff
@@ -0,0 +1,179 @@
+
+	client dynamic.Interface
+	mapper meta.RESTMapper
+}
+
+// EnsureManifestsCreated ensures that all resource manifests from the specified directory are created.
+// This function will keep retrying creation until no errors are reported. After specified timeout, the last set
+// of observed errors are returned to caller.
+func EnsureManifestsCreated(manifestDir string, restConfig *rest.Config, ctx context.Context, options CreateOptions) error {
+	var err error
+
+	options.client, err = dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	gr, err := restmapper.GetAPIGroupResources(dc)
+	if err != nil {
+		return err
+	}
+
+	options.mapper = restmapper.NewDiscoveryRESTMapper(gr)
+
+	manifests, err := load(manifestDir, options)
+	if err != nil {
+		return err
+	}
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var lastCreateError error
+	err = wait.PollImmediateUntil(500*time.Millisecond, func() (bool, error) {
+		err := create(manifests, options)
+		if err != nil {
+			return true, nil
+		}
+		lastCreateError = err
+		return false, nil
+	}, ctx.Done())
+
+	// Return the last observed set of errors from the create process instead of timeout error.
+	if lastCreateError != nil {
+		return lastCreateError
+	}
+
+	return err
+}
+
+func create(manifests map[string]*unstructured.Unstructured, options CreateOptions) error {
+	// Sort all manifests, so in case they use number prefixes, we follow their order, which will increase
+	// the chances for the create loop to succeed faster.
+	var (
+		sortedManifestPaths []string
+	)
+	for key := range manifests {
+		sortedManifestPaths = append(sortedManifestPaths, key)
+	}
+	sort.Strings(sortedManifestPaths)
+
+	errs := map[string][]error{}
+
+	for _, path := range sortedManifestPaths {
+		gvk := manifests[path].GetObjectKind().GroupVersionKind()
+		mappings, err := options.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			errs[path] = append(errs[path], fmt.Errorf("unable to get REST mapping: %v", err))
+		}
+		if mappings.Scope.Name() == meta.RESTScopeNameRoot {
+			_, err = options.client.Resource(mappings.Resource).Create(manifests[path], metav1.CreateOptions{})
+		} else {
+			_, err = options.client.Resource(mappings.Resource).Namespace(manifests[path].GetNamespace()).Create(manifests[path], metav1.CreateOptions{})
+		}
+
+		// Resource already exists means we already succeeded
+		// This should not happen as we remove already created items from the manifest list.
+		if kerrors.IsAlreadyExists(err) {
+			continue
+		}
+
+		if err != nil {
+			errs[path] = append(errs[path], err)
+			continue
+		}
+
+		// Creation succeed lets remove the manifest from the list to avoid creating it second time
+		delete(manifests, path)
+	}
+
+	return reportErrors(errs)
+}
+
+func reportErrors(errors map[string][]error) error {
```

Comment:
report? Didn't we have an aggregated error formatter in library-go already?

Reply:
this is special, it use paths

Reply:
ok, then it's fine